pin/gc/gc.go

```diff
@@ -35,144 +54,159 @@ type Result struct {
+			if err != nil {
+				output <- Result{Error: err}
+				criticalError = ErrCannotFetchAllLinks
+			}
+			if finished {
+				break
+			}
+		}
+		if criticalError != nil {
 			return
 		}
-		emark.Append(logging.LoggableMap{
-			"blackSetSize": fmt.Sprintf("%d", gcs.Len()),
-		})
-		emark.Done()
-		esweep := log.EventBegin(ctx, "GC.sweep")
 
+		// Add white objects
 		keychan, err := bs.AllKeysChan(ctx)
 		if err != nil {
 			output <- Result{Error: err}
 			return
 		}
 
-		errors := false
-		var removed uint64
-
 	loop:
 		for {
 			select {
-			case k, ok := <-keychan:
+			case c, ok := <-keychan:
 				if !ok {
 					break loop
 				}
-				if !gcs.Has(k) {
-					err := bs.DeleteBlock(k)
-					removed++
-					if err != nil {
-						errors = true
-						output <- Result{Error: &CannotDeleteBlockError{k, err}}
-						//log.Errorf("Error removing key from blockstore: %s", err)
-						// continue as error is non-fatal
-						continue loop
-					}
-					select {
-					case output <- Result{KeyRemoved: k}:
-					case <-ctx.Done():
-						break loop
-					}
-				}
+				tri.InsertFresh(c)
 			case <-ctx.Done():
-				break loop
+				output <- Result{Error: ctx.Err()}
+				return
 			}
 		}
-		esweep.Append(logging.LoggableMap{
-			"whiteSetSize": fmt.Sprintf("%d", removed),
-		})
-		esweep.Done()
-		if errors {
-			output <- Result{Error: ErrCannotDeleteSomeBlocks}
-		}
-	}()
 
-	return output
-}
+		// Regain lock
+		unlocker, elock := getGCLock(ctx, bs)
 
-func Descendants(ctx context.Context, getLinks dag.GetLinks, set *cid.Set, roots []*cid.Cid) error {
-	for _, c := range roots {
-		set.Add(c)
+		defer unlocker.Unlock()
+		defer elock.Done()
 
-		// EnumerateChildren recursively walks the dag and adds the keys to the given set
-		err := dag.EnumerateChildren(ctx, getLinks, c, set.Visit)
+		err = pn.Flush()
 		if err != nil {
-			return err
+			output <- Result{Error: err}
+			return
 		}
-	}
+		// Add the roots again, they might have changed
+		addRoots(tri, pn, bestEffortRoots)
 
-	return nil
-}
+		// This prevents incremental and concurrent GCing
+		for _, v := range pn.DirectKeys() {
+			tri.blacken(v, enumStrict)
+		}
+		for _, v := range pn.InternalPins() {
+			tri.InsertGray(v, true)
+		}
 
-// ColoredSet computes the set of nodes in the graph that are pinned by the
-// pins in the given pinner.
-func ColoredSet(ctx context.Context, pn pin.Pinner, ls dag.LinkService, bestEffortRoots []*cid.Cid, output chan<- Result) (*cid.Set, error) {
-	// KeySet currently implemented in memory, in the future, may be bloom filter or
-	// disk backed to conserve memory.
-	errors := false
-	gcs := cid.NewSet()
-	getLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
-		if err != nil {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+		// Reenumerate, fast as most will be duplicate
+		for {
+			finished, err := tri.EnumerateStep(ctx, getLinks, bestEffortGetLinks)
```

Comment:
shouldnt we technically loop back up to the top of the function? technically you're supposed to just keep doing all that until you end up with an empty gray set, no?

Reply:
I don't understand, Line 156 loops this until finished is true L162.

Reply:
nvm, i see that we do loop up above.